File: apps/tabtin_django/tabtin/sentry.py
"""Sentry 错误监控接入（errors-only）。

启用条件：env 注入 ``SENTRY_DSN=https://...``（不注入时不初始化 → 零开销，
与 OTel 接入同款 opt-in 模式）。sentry-sdk 未安装时同样静默降级。

字段契约（tags 白名单 / 脱敏红线 / 各端取值来源）见
``docs/agent/error-context-schema.md``——本模块只允许上报契约里登记过的字段。

三个接入点：
1. ``init_sentry()``            —— wsgi / asgi / celery 三个进程入口调用
   （管理命令与 ``scripts/*.py`` 独立脚本不初始化——错误监控只覆盖常驻服务）；
2. ``before_send``              —— 全局事件钩子：注入 ContextVar 业务 tags + 脱敏 + 同指纹限频；
3. ``capture_api_exception()``  —— ninja 兜底 handler 显式上报（ninja 会把未捕获
   异常转成 500 响应吞掉，不发 ``got_request_exception`` 信号，Django 集成抓不到）。
"""

from __future__ import annotations

import logging
import os
import re
import time
from typing import Any, Optional

logger = logging.getLogger(__name__)

_initialized = False

# ── tags 白名单（跨端契约，新字段先改 error-context-schema.md 再登记到这里） ──

#: ContextVar getter → tag 键 的映射。取不到值（None/空）的键不上报。
_THREAD_CONTEXT_TAGS = (
    ("organization_id", "get_current_organization_id"),
    ("space_id", "get_current_space_id"),
    ("agent_id", "get_current_execution_agent_id"),
)
_PLATFORM_CONTEXT_TAGS = (
    ("session_id", "get_current_session_id"),
    ("run_id", "get_current_run_id"),
)

# ── 脱敏（结构性现场原则：token / 手机号 / 邮箱 / 家目录用户名不出境） ──

_SCRUB_RULES: tuple[tuple[re.Pattern[str], Any], ...] = (
    (re.compile(r"(bearer\s+)[A-Za-z0-9\-._~+/]{8,}=*", re.IGNORECASE), r"\1<redacted>"),
    (
        re.compile(
            r"(\"?(?:password|passwd|pwd|secret|token|api[_-]?key|access[_-]?token"
            r"|refresh[_-]?token|authorization|cookie|set-cookie)\"?\s*[:=]\s*\"?)"
            r"([^\s\"',}]{4,})",
            re.IGNORECASE,
        ),
        r"\1<redacted>",
    ),
    (re.compile(r"\beyJ[A-Za-z0-9\-_]+\.[A-Za-z0-9\-_]+\.[A-Za-z0-9\-_]+"), "<redacted-jwt>"),
    (
        re.compile(r"([A-Za-z0-9._%+-]+)@([A-Za-z0-9.-]+\.[A-Za-z]{2,})"),
        lambda m: f"{m.group(1)[:1]}***@{m.group(2)}",
    ),
    (
        re.compile(r"\b1[3-9]\d{9}\b"),
        lambda m: f"{m.group(0)[:3]}****{m.group(0)[7:]}",
    ),
    (re.compile(r"(/Users/)[^/\s]+"), r"\1<user>"),
    (re.compile(r"(/home/)[^/\s]+"), r"\1<user>"),
    (re.compile(r"([A-Za-z]:\\Users\\)[^\\/\s]+"), r"\1<user>"),
)


def scrub_text(text: str) -> str:
    """对一段文本做脱敏（与 Electron 诊断包 ``diagnostics-redact.ts`` 同款规则）。"""
    if not isinstance(text, str) or not text:
        return text or ""
    for pattern, replacement in _SCRUB_RULES:
        text = pattern.sub(replacement, text)
    return text


# ── 同指纹限频（洪水保护）────────────────────────────────────────────
# 仓库里已有高频 capture_message 调用点（如 relay silent-drop，dogfood 实测
# 每秒可达上百条），DSN 一配置就会被激活；SDK 自带的 dedupe 只拦完全相同的
# 异常对象。这里按消息/异常指纹做进程内限频兜底：同指纹每窗口最多放行
# _RATE_LIMIT_MAX 条，超出的在 before_send 里直接丢弃。

_RATE_LIMIT_WINDOW_S = 60.0
_RATE_LIMIT_MAX = 5
_RATE_LIMIT_TABLE_CAP = 512

#: fingerprint -> (窗口起点时间戳, 窗口内已放行条数)
_fingerprint_hits: dict[str, tuple[float, int]] = {}


def _event_fingerprint(event: dict) -> str:
    exception = event.get("exception")
    if isinstance(exception, dict):
        values = exception.get("values") or []
        if values and isinstance(values[0], dict):
            first = values[0]
            return f"exc:{first.get('type')}:{str(first.get('value'))[:120]}"
    logentry = event.get("logentry")
    if isinstance(logentry, dict) and logentry.get("message"):
        return f"log:{str(logentry['message'])[:160]}"
    return f"msg:{str(event.get('message'))[:160]}"


def _rate_limited(event: dict, now: Optional[float] = None) -> bool:
    """同指纹限频判定。返回 True 表示该事件应被丢弃。"""
    now = time.monotonic() if now is None else now
    key = _event_fingerprint(event)
    started, count = _fingerprint_hits.get(key, (0.0, 0))
    if now - started >= _RATE_LIMIT_WINDOW_S:
        started, count = now, 0
    if count >= _RATE_LIMIT_MAX:
        _fingerprint_hits[key] = (started, count)
        return True
    if len(_fingerprint_hits) >= _RATE_LIMIT_TABLE_CAP and key not in _fingerprint_hits:
        # 表满先清过期项；仍满则整表重置（宁可放行也不无界增长）
        expired = [k for k, (ts, _) in _fingerprint_hits.items() if now - ts >= _RATE_LIMIT_WINDOW_S]
        for k in expired:
            del _fingerprint_hits[k]
        if len(_fingerprint_hits) >= _RATE_LIMIT_TABLE_CAP:
            _fingerprint_hits.clear()
    _fingerprint_hits[key] = (started, count + 1)
    return False


def collect_context_tags() -> dict[str, str]:
    """从平台 ContextVar 收集当前执行现场的业务 tags（白名单内、非空才带）。

    Agent 执行链路的 middleware（AgentRunContextMiddleware 等）与 runtime
    会填充这些 ContextVar；普通 API 请求大多为空——空值不上报，不算异常。
    """
    tags: dict[str, str] = {}
    try:
        from apps.services.common import thread_context

        for tag_key, getter_name in _THREAD_CONTEXT_TAGS:
            value = getattr(thread_context, getter_name)()
            if value:
                tags[tag_key] = str(value)
    except Exception:  # ContextVar 读取绝不能反过来弄坏错误上报
        logger.debug("[Sentry] thread_context tags 收集失败", exc_info=True)
    try:
        from apps.services.common import platform_context

        for tag_key, getter_name in _PLATFORM_CONTEXT_TAGS:
            value = getattr(platform_context, getter_name)()
            if value:
                tags[tag_key] = str(value)
    except Exception:
        logger.debug("[Sentry] platform_context tags 收集失败", exc_info=True)
    return tags


def scrub_event(event: dict, _hint: Optional[dict] = None) -> Optional[dict]:
    """``before_send`` 钩子：同指纹限频 + 注入业务 tags + 全事件脱敏。

    返回 None 表示丢弃事件（限频命中）。除 ContextVar / 限频表外为纯函数，
    单测直接构造 event dict 验证。
    """
    if _rate_limited(event):
        return None

    tags = event.setdefault("tags", {})
    for key, value in collect_context_tags().items():
        tags.setdefault(key, value)

    # message / logentry
    if isinstance(event.get("message"), str):
        event["message"] = scrub_text(event["message"])
    logentry = event.get("logentry")
    if isinstance(logentry, dict) and isinstance(logentry.get("message"), str):
        logentry["message"] = scrub_text(logentry["message"])

    # exception values
    exception = event.get("exception")
    if isinstance(exception, dict):
        for value in exception.get("values") or []:
            if isinstance(value, dict) and isinstance(value.get("value"), str):
                value["value"] = scrub_text(value["value"])

    # breadcrumbs（Sentry 集成会把 logging 面包屑带进来）——message 和
    # data 里的字符串值都要脱敏（http 面包屑的 url query 可能带 token/邮箱）
    breadcrumbs = event.get("breadcrumbs")
    crumb_list = breadcrumbs.get("values") if isinstance(breadcrumbs, dict) else breadcrumbs
    if isinstance(crumb_list, list):
        for crumb in crumb_list:
            if not isinstance(crumb, dict):
                continue
            if isinstance(crumb.get("message"), str):
                crumb["message"] = scrub_text(crumb["message"])
            data = crumb.get("data")
            if isinstance(data, dict):
                for key, value in data.items():
                    if isinstance(value, str):
                        data[key] = scrub_text(value)

    # 请求体不出境（结构性现场原则）；url/headers 由 send_default_pii=False 兜底
    request = event.get("request")
    if isinstance(request, dict):
        request.pop("data", None)
        if isinstance(request.get("query_string"), str):
            request["query_string"] = scrub_text(request["query_string"])

    return event


def resolve_sentry_environment() -> str:
    """Return the governed observability environment.

    ``test`` remains readable for already-running legacy deployments, but new
    test deployments must use ``test-new`` so their signal is isolated from
    the historical pool.
    """
    configured = os.getenv("SENTRY_ENVIRONMENT", "").strip()
    if configured:
        if configured not in {"test", "test-new", "production"}:
            raise ValueError(
                "SENTRY_ENVIRONMENT must be 'test', 'test-new', or 'production'"
            )
        return configured
    return "test-new" if os.getenv("DEBUG", "False").lower() == "true" else "production"


def resolve_sentry_release() -> str | None:
    """事件必须能回到镜像源码；显式 release 仅作为兼容覆盖。"""
    configured = os.getenv("SENTRY_RELEASE", "").strip()
    if configured:
        return configured
    source_sha = os.getenv("MUSE_SOURCE_SHA", os.getenv("MUSE_GIT_SHA", "")).strip()
    return f"tabtin-django@{source_sha}" if source_sha else None


def resolve_sentry_initial_scope() -> dict[str, dict[str, str]]:
    """为每个 Django 事件保留实际镜像源码 SHA，供告警归因使用。"""
    source_sha = os.getenv("MUSE_SOURCE_SHA", os.getenv("MUSE_GIT_SHA", "")).strip()
    return {"tags": {"source_sha": source_sha}} if source_sha else {}


def resolve_sentry_release() -> str | None:
    """事件必须能回到镜像源码；显式 release 仅作为兼容覆盖。"""
    configured = os.getenv("SENTRY_RELEASE", "").strip()
    if configured:
        return configured
    source_sha = os.getenv("MUSE_SOURCE_SHA", os.getenv("MUSE_GIT_SHA", "")).strip()
    return f"tabtin-django@{source_sha}" if source_sha else None


def resolve_sentry_initial_scope() -> dict[str, dict[str, str]]:
    """为每个 Django 事件保留实际镜像源码 SHA，供告警归因使用。"""
    source_sha = os.getenv("MUSE_SOURCE_SHA", os.getenv("MUSE_GIT_SHA", "")).strip()
    return {"tags": {"source_sha": source_sha}} if source_sha else {}


def init_sentry() -> bool:
    """按 env 初始化 sentry-sdk。返回是否真正启用（供自检/测试断言）。"""
    global _initialized
    dsn = os.getenv("SENTRY_DSN", "").strip()
    if not dsn:
        return False
    from tabtin.startup_policy import (
        StartupCapability,
        resolve_endpoint_setting,
        resolve_startup_policy,
    )

    policy = resolve_startup_policy(os.environ)
    if not policy.allows(
        StartupCapability.SENTRY,
        explicitly_configured=True,
    ):
        return False
    try:
        resolve_endpoint_setting(
            os.environ,
            "SENTRY_DSN",
            saas_default="",
        )
    except ValueError:
        logger.warning("[Sentry] Community company endpoint blocked; error reporting disabled")
        return False
    try:
        import sentry_sdk
        from sentry_sdk.integrations.celery import CeleryIntegration
        from sentry_sdk.integrations.django import DjangoIntegration
    except ImportError:
        logger.warning("[Sentry] SENTRY_DSN 已配置但 sentry-sdk 未安装，错误上报未启用")
        return False

    try:
        environment = resolve_sentry_environment()
    except ValueError as error:
        logger.error("[Sentry] %s；错误上报未启用", error)
        return False
    release = resolve_sentry_release()

    sentry_sdk.init(
        dsn=dsn,
        environment=environment,
        release=release,
        integrations=[DjangoIntegration(), CeleryIntegration()],
        # errors-only：不采集性能事务（自部署也按 errors-only 模式跑）
        traces_sample_rate=0,
        # 结构性现场原则：不自动带 PII（cookies / auth headers / user ip）
        send_default_pii=False,
        # 结构性现场原则：栈帧局部变量常含用户消息/文档内容（内容性现场），
        # 且 SDK 默认 scrubber 只按变量名打码、拦不住值——整体关闭。
        include_local_variables=False,
        before_send=scrub_event,
    )
    # ``initial_scope`` is not accepted by every supported sentry-sdk 2.x
    # release. Set the deployment tag through the long-standing public API
    # after initialization instead, so an observability setting can never
    # prevent the API process from starting.
    for key, value in resolve_sentry_initial_scope().get("tags", {}).items():
        sentry_sdk.set_tag(key, value)
    _initialized = True
    logger.info("[Sentry] 已启用 environment=%s release=%s", environment, release or "<unset>")
    return True


def is_enabled() -> bool:
    return _initialized


def capture_api_exception(request: Any, exc: Exception) -> None:
    """在 ninja 兜底异常 handler 里显式上报未捕获异常。

    ninja 把异常转成 500 JSON 后 Django 不再抛出，``got_request_exception``
    信号不触发，Sentry Django 集成抓不到——必须在 handler 里显式 capture。
    user 从 ``request.auth``（JWTAuth 放置的 User 实例）取，只带内部 ID 与
    昵称（契约：绝不上报手机号/邮箱）。
    """
    if not _initialized:
        return
    try:
        import sentry_sdk

        with sentry_sdk.isolation_scope() as scope:
            auth_user = getattr(request, "auth", None)
            user_id = getattr(auth_user, "id", None)
            if user_id is not None:
                scope.set_user(
                    {
                        "id": str(user_id),
                        "username": str(getattr(auth_user, "nickname", "") or ""),
                    }
                )
            scope.set_tag("handled_by", "ninja_generic_handler")
            sentry_sdk.capture_exception(exc)
    except Exception:
        logger.debug("[Sentry] capture_api_exception 失败", exc_info=True)

File: apps/tabtin-daemon/src/platform/observability/logging/sentry.ts
/**
 * Daemon Sentry 错误监控接入（，errors-only）。
 *
 * 启用条件：`SENTRY_DSN` 环境变量或 config.json 的 `sentry_dsn` 字段有值。
 * Daemon 不读仓库根 .env（部署形态是 npm 全局安装 + systemd/launchd），
 * DSN 走运维口径：`Environment=SENTRY_DSN=...` 或
 * `tabtin-daemon config --set sentry_dsn=...`。不配置时不加载 SDK，零开销。
 *
 * 字段契约（tags 白名单 / 脱敏红线）：docs/agent/error-context-schema.md。
 * 脱敏与 Electron 同源（@tabtin/shared/sentry-scrub）；同指纹限频与 Django
 * 端 muse/sentry.py 同口径（每指纹每分钟最多 5 条），防止 relay/重试类
 * 高频错误路径打爆自部署 Sentry。
 *
 * Runtime（@tabtin/agent-runtime）本身不依赖 @sentry/*：run 级致命错误统一
 * 由宿主 DaemonAgentHost.handleQuery 的 catch 调 `captureRunError` 收口，
 * capture 时显式带 run 上下文 tags——Daemon 并发多 session，全局 scope
 * setTag 会互相污染，收口处显式传 tags 才并发安全。
 */

import { createRequire } from 'node:module';
import { scrubSentryEvent } from '@tabtin/shared/sentry-scrub';
import type { DaemonConfig } from '../../../base/types/daemon-config.js';
import { readDaemonVersion } from '../../../platform/system/update/daemon-version.js';
import type { Logger } from './logger.js';

type SentryNode = typeof import('@sentry/node');

const requireModule = createRequire(import.meta.url);

let sdk: SentryNode | null = null;

/** 供诊断/自检读取：Daemon Sentry 是否真正启用。 */
export function isSentryEnabled(): boolean {
  return sdk !== null;
}

// ── 同指纹限频（洪水保护，与 Django muse/sentry.py 同口径） ──────────

const RATE_LIMIT_WINDOW_MS = 60_000;
const RATE_LIMIT_MAX = 5;
const RATE_LIMIT_TABLE_CAP = 512;

interface SentryEventLike {
  message?: unknown;
  logentry?: { message?: unknown };
  exception?: { values?: Array<{ type?: unknown; value?: unknown }> };
}

/** fingerprint -> [窗口起点, 窗口内已放行条数] */
const fingerprintHits = new Map<string, [number, number]>();

function eventFingerprint(event: SentryEventLike): string {
  const first = event.exception?.values?.[0];
  if (first) {
    return `exc:${String(first.type)}:${String(first.value).slice(0, 120)}`;
  }
  if (typeof event.logentry?.message === 'string') {
    return `log:${event.logentry.message.slice(0, 160)}`;
  }
  return `msg:${String(event.message).slice(0, 160)}`;
}

/**
 * 同指纹限频判定。返回 true 表示该事件应被丢弃。导出仅为单测。
 * 用单调钟（performance.now，与 Django 版 time.monotonic 对齐）——挂钟回拨
 * 会让 `now - started < 0`、窗口迟迟不重置，限频窗口被无声拉长。
 */
export function isRateLimited(event: SentryEventLike, now = performance.now()): boolean {
  const key = eventFingerprint(event);
  let [started, count] = fingerprintHits.get(key) ?? [0, 0];
  if (now - started >= RATE_LIMIT_WINDOW_MS) {
    started = now;
    count = 0;
  }
  if (count >= RATE_LIMIT_MAX) {
    fingerprintHits.set(key, [started, count]);
    return true;
  }
  if (fingerprintHits.size >= RATE_LIMIT_TABLE_CAP && !fingerprintHits.has(key)) {
    // 表满先清过期项；仍满则整表重置（宁可放行也不无界增长）
    for (const [k, [ts]] of fingerprintHits) {
      if (now - ts >= RATE_LIMIT_WINDOW_MS) fingerprintHits.delete(k);
    }
    if (fingerprintHits.size >= RATE_LIMIT_TABLE_CAP) fingerprintHits.clear();
  }
  fingerprintHits.set(key, [started, count + 1]);
  return false;
}

/** 单测隔离用：清空限频表。 */
export function resetRateLimiter(): void {
  fingerprintHits.clear();
}

// ── 初始化 ──────────────────────────────────────────────────────────

/**
 * 初始化 Daemon Sentry。在 createDaemonContainer 里 config 加载后立即调用
 * （早于所有服务构建，保证后续任何位置的 captureRunError 可用）。
 * 失败绝不阻塞启动。
 */
export function initSentryDaemon(config: DaemonConfig, logger: Logger): void {
  const dsn = (process.env.SENTRY_DSN || config.sentry_dsn || '').trim();
  if (!dsn) return;

  try {
    // 同步 require（CJS 入口）：与 Electron 主进程同款条件加载手法，
    // 未配置 DSN 时完全不加载 SDK。
    const Sentry = requireModule('@sentry/node') as SentryNode;
    Sentry.init({
      dsn,
      environment: (process.env.SENTRY_ENVIRONMENT || '').trim() || 'prod',
      release: `tabtin-daemon@${readDaemonVersion()}`,
      // errors-only：不采集性能事务
      tracesSampleRate: 0,
      sendDefaultPii: false,
      // 进程级钩子的唯一真相在 ProcessManager（captureFatal + flush 后再优雅
      // 停机）。SDK 默认的 OnUncaughtException / OnUnhandledRejection 集成
      // 会：① 对同一崩溃重复上报一条；② OnUncaughtException 在检测到其他
      // handler 后仍可能自行 exit，与 ProcessManager 的 graceful shutdown
      // （15s 收尾）打架。两个都禁掉，保证单一收口。
      integrations: (defaults) =>
        defaults.filter(
          (i) => i.name !== 'OnUncaughtException' && i.name !== 'OnUnhandledRejection',
        ),
      beforeSend: (event) => {
        if (isRateLimited(event as SentryEventLike)) return null;
        return scrubSentryEvent(event);
      },
    });

    // 静态 tags：Daemon 是单 organization / 单 device 模型，进程级设置即可
    Sentry.setTag('device_id', config.device_id);
    Sentry.setTag('organization_id', config.organization_id);

    sdk = Sentry;
    logger.info(`[Sentry] Daemon 错误上报已启用 environment=${process.env.SENTRY_ENVIRONMENT || 'prod'}`);
  } catch (error) {
    logger.error('[Sentry] Daemon 初始化失败（继续启动，不上报）:', error);
  }
}

// ── 上报入口 ────────────────────────────────────────────────────────

/** run 错误上下文（键名以 error-context-schema.md 白名单为准）。 */
export interface RunErrorContext {
  handled_by: string;
  error_category: 'AGENT_RUN_FATAL' | 'AGENT_DOOM_LOOP' | 'AGENT_PROTOCOL_FATAL';
  error_code: string;
  run_id?: string;
  session_id?: string;
  agent_id?: string;
  organization_id?: string;
  workspace_id?: string;
  space_id?: string;
  task_id?: string;
}

/**
 * Run 级致命错误收口上报。未启用时是 no-op。
 * 低基数分类走 tags，高基数关联 ID 走 contexts.tabtin；两者都用 withScope
 * 逐事件设置，避免并发 session 互相污染。
 */
export function captureRunError(error: unknown, context: RunErrorContext): void {
  if (!sdk) return;
  try {
    sdk.withScope((scope) => {
      scope.setTag('handled_by', context.handled_by);
      scope.setTag('error_category', context.error_category);
      scope.setTag('error_code', context.error_code);
      const tabtinContext = Object.fromEntries(Object.entries({
        organization_id: context.organization_id,
        workspace_id: context.workspace_id,
        space_id: context.space_id,
        agent_id: context.agent_id,
        session_id: context.session_id,
        run_id: context.run_id,
        task_id: context.task_id,
      }).filter(([, value]) => Boolean(value)));
      scope.setContext('muse', tabtinContext);
      scope.setFingerprint([
        'agent-run',
        context.error_category,
        context.error_code,
      ]);
      sdk!.captureException(error);
    });
  } catch {
    // 上报失败不影响主流程
  }
}

/**
 * 进程级兜底上报（uncaughtException / 致命 unhandledRejection）。
 * 返回的 Promise **保证 settle**（flush 2s 超时 + 3s watchdog race 双保险），
 * 调用方 `.finally(shutdown)` 不会因 SDK flush 违约悬挂而永不进入 shutdown。
 */
export function captureFatal(error: unknown, kind: string): Promise<void> {
  if (!sdk) return Promise.resolve();
  try {
    sdk.withScope((scope) => {
      scope.setTag('handled_by', kind);
      sdk!.captureException(error);
    });
    const flushed = sdk
      .flush(2_000)
      .then(() => undefined)
      .catch(() => undefined);
    const watchdog = new Promise<void>((resolve) => {
      const timer = setTimeout(resolve, 3_000);
      timer.unref?.();
    });
    return Promise.race([flushed, watchdog]);
  } catch {
    return Promise.resolve();
  }
}
